cpp/src/arrow/type.h

```diff
@@ -116,7 +116,7 @@ struct DataType {
 
   bool Equals(const DataType* other) {
     // Call with a pointer so more friendly to subclasses
-    return this == other || (this->type == other->type);
+    return this == other || ((other != nullptr) && (this->type == other->type));
```

Comment:
Interesting note (a very knowledgeable C++ pointed this out to me):
```
DataType* t = nullptr
assert_true(t->Equals(nullptr))
```
will work work (and not segfault).  I believe your suggestion will return false.  Of course once we get to that point we might want undefined behavior anyways.